Load initial data from XCTS solver in GH execs

nilsvu opened the pull request:
## Proposed changes

This PR connects the XCTS elliptic solver with the GH execs:

- Output lapse, shift, spatial metric and extrinsic curvature from the XCTS solver.
- Load these variables in the GH execs and transform to GH variables.

I tested this with a Kerr solution. It should also work with BBH initial data, but I haven't done much testing yet. Here are some example input files: https://github.com/nilsvu/spectre/commit/78c936be17f4eb59f3d039cefb2687967c2a1072

Current limitations:

- No interpolation. Run the initial data solver on the same grid as the evolution.
- (fixed in #3824) ~Use the last iteration number of the initial data solver as the `ObservationValue` in the GH input file. This should get smoothed out of course.~
- dt_lapse and dt_shift are both chosen to be zero in the "inertial" frame at the moment. Which options do we want to support here?

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Support loading a subset of tags from file

nilsvu committed:
Add gr::time_derivative_of_spatial_metric

nilsvu committed:
Load initial data from XCTS solver in GH execs

nilsvu commented:
@geoffrey4444 could you give this a brief look when you get the chance, to see if it goes in the direction you want?

knelli2 commented:
Question: Does this change how we read in SpEC initial data? I imagine we will want to support for this feature for a while still.

nilsvu commented:
@knelli2 That doesn't change. Set `Variables: GeneralizedHarmonic` to load spacetime metric, phi, pi (current behavior). Or set `Variables: Adm` to load spatial metric, lapse, shift, extrinsic curvature (new feature).

nilsvu committed:
Add a missing include

knelli2 commented on src/PointwiseFunctions/GeneralRelativity/TimeDerivativeOfSpatialMetric.cpp:
```diff
@@ -0,0 +1,90 @@
+    const tnsr::ijj<DataType, SpatialDim, Frame>& deriv_spatial_metric,
+    const tnsr::ii<DataType, SpatialDim, Frame>& extrinsic_curvature) {
+  destructive_resize_components(dt_spatial_metric, get_size(get(lapse)));
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = 0; j <= i; ++j) {
```
Optional/Question: Now that tensor expressions are working, maybe we should consider starting to use those in this and future PRs. Thoughts? (I don't know how much work it is switching from loops to TEs)

nilsvu replied:
Yea I want to do that ASAP, though my understanding is that a few performance PRs are not in yet. Let's wait for the go-ahead by @macedo22 :)

macedo22 replied:
Yes, that's correct. I will get the PR(s) with these changes in ASAP!

knelli2 commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Actions/Test_NumericInitialData.cpp:
```diff
@@ -0,0 +1,244 @@
+          selected_fields) {
+    const auto selected_vars =
+        get<Tags::NumericInitialDataVariables<TestOptionGroup>>(cache);
+    if (selected_vars == NumericInitialDataVariables::GeneralizedHarmonic) {
+      CHECK(get<importers::Tags::Selected<
+                gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>>>(
```
Also `CHECK_FALSE` that the adm variables were selected here. Vice-versa below.

knelli2 commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Actions/Test_NumericInitialData.cpp:
```diff
@@ -0,0 +1,244 @@
+  // SetNumericInitialData
+  ActionTesting::next_action<element_array>(make_not_null(&runner), element_id);
+
+  // Check result
+  Approx custom_approx = Approx::custom().epsilon(1.e-3).scale(1.0);
```
Why so large? Add a comment clarifying please.

knelli2 commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Actions/Test_NumericInitialData.cpp:
```diff
@@ -0,0 +1,244 @@
+  test_numeric_initial_data(
+      GeneralizedHarmonic::NumericInitialDataVariables::Adm);
+}
+
+}  // namespace GeneralizedHarmonic
```
Arbitrary line: Add an error test where you fail to create a `GeneralizedHarmonic::NumericInitialDataVariables` from options because of a bad input name like:
```
Variables: KyleAlwaysMisspellsThngs
```

nilsvu replied:
I believe this is already sufficiently tested as part of the option parsing, don't you agree?

knelli2 replied:
Yes it is.

nilsvu commented:
@knelli2 Thanks for your review. I pushed a fixup to your comments, but I also changed the `NumericInitialData.hpp` file quite significantly to support specifying the dataset name in the H5 file that each tensor gets loaded from. I forgot to keep those changes in a fixup commit, sorry about that! Only the `NumericInitialData.hpp` and `ReadVolumeData.hpp` files changed, though.

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Actions/NumericInitialData.hpp:
```diff
@@ -0,0 +1,325 @@
+ * - gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>
+ * - GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>
+ * - GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial>
+ * - ::Tags::Time: To compute numerical derivatives on time-dependent domains we
+ *   have to temporarily set the time to the initial time, so it gets picked up
+ *   by compute items for the Jacobian. We set the time back to NaN before this
+ *   action completes.
```
@wthrowe I had to put in this workaround for an issue that @geoffrey4444 encountered with time-dependent domains. Essentially, `::Tags::Time` is NaN during initialization, but for taking numeric derivatives of the initial data we have to set it to the `InitialTime` temporarily. When you get the chance, could we try to come up with a more sustainable solution? Happy to discuss in a call some time.

wthrowe replied:
Just set it where it's first created.

nilsvu replied:
So you mean, set `::Tags::Time` to `InitialTime` in `Initialization::Actions::TimeAndTimeStep`?

wthrowe replied:
Use `initial_time.value()` rather than the input parameter, but otherwise yes.  (They should be the same, but this is consistent with what is done elsewhere.)

knelli2 commented on src/Evolution/Systems/GeneralizedHarmonic/Actions/NumericInitialData.hpp:
```diff
@@ -0,0 +1,325 @@
+
+// Collect all variables that we support loading from volume data files
+using all_numeric_vars =
+    tmpl::remove_duplicates<tmpl::append<adm_vars, gh_vars>>;
```
Optional: There aren't really any duplicates between the ADM vars and the GH vars right now. In the future, if we are going to add to this list to support more initial data options then it's ok, but is it necessary now?

nilsvu replied:
ok, removed this (modified the fixup) and rebased

geoffrey4444 reviewed (dismissed):
I tested this on ocean, and I was able to load initial data from SolveXCTS and evolve it with EvolveGhBinaryBlackHole

nilsvu committed:
Set ::Tags::Time to initial time during initialization

This is needed so Jacobians of time-dependent domains can
be computed to take numerical derivatives of initial data.

nilsvu commented:
@knelli2 Could you check if you're happy with the fixup?

knelli2 commented:
What's been updated since the last time I reviewed this?

nilsvu commented:
I think I only rebased and squashed the fixup.

Edit: Oh I added comments to `BinaryBlackHole.yaml` while rebasing.

knelli2 reviewed (dismissed):
Ok looks good.

nilsvu commented:
Oh you probably referred to the "updated" label. I just figured I put the label instead of pinging people, but that probably wasn't a great idea :D

@wthrowe could you provide a cursory review and get this toward merger? @knelli2 and @geoffrey4444 have already reviewed.

wthrowe commented on src/IO/Importers/Actions/ReadVolumeData.hpp:
```diff
@@ -44,6 +46,24 @@ struct ReadAllVolumeDataAndDistribute;
+ * with the name of the dataset in the volume data file.
+ *
+ * Set the value to a dataset name to import the `FieldTag` from that dataset,
+ * or to `std::nullopt` to skip importing the `FieldTag. The dataset name
```
Missing backtick.

nilsvu replied:
fixed, thanks 👍

wthrowe reviewed (changes requested):
Read the documentation, mostly skimmed the rest.  One part didn't render correctly.
